packages/yarnpkg-pnp/sources/loader/hydrateRuntimeState.ts

```diff
@@ -14,13 +14,26 @@ export function hydrateRuntimeState(data: SerializedState, {basePath}: HydrateRu
     ? new RegExp(data.ignorePatternData)
     : null;
 
+  const packageLocatorsByLocations = new Map<PortablePath, PhysicalPackageLocator | null>();
+  const packageLocationLengths = new Set<number>();
+
   const packageRegistry = new Map(data.packageRegistryData.map(([packageName, packageStoreData]) => {
     return [packageName, new Map(packageStoreData.map(([packageReference, packageInformationData]) => {
+      if ((packageName === null) !== (packageReference === null))
+        throw new Error(`Assertion failed: The name and reference should be null, or neither should`);
+
+      if (!packageInformationData.discardFromLookup) {
+        // @ts-expect-error: TypeScript isn't smart enough to understand the type assertion
+        const packageLocator: PhysicalPackageLocator = {name: packageName, reference: packageReference};
+        packageLocatorsByLocations.set(packageInformationData.packageLocation, packageLocator);
+
+        packageLocationLengths.add(packageInformationData.packageLocation.length);
+      }
+
       return [packageReference, {
-        // We use ppath.join instead of ppath.resolve because:
-        // 1) packageInformationData.packageLocation is a relative path when part of the SerializedState
-        // 2) ppath.join preserves trailing slashes
-        packageLocation: ppath.join(absolutePortablePath, packageInformationData.packageLocation),
+        // The packageLocation is relative and starts with `./`
+        // inlined because this is more performant than `ppath.join`
+        packageLocation: `${absolutePortablePath}/${packageInformationData.packageLocation.slice(2)}`,
```

Comment:
It doesn't necessarily starts with `./` when the path is outside the local project (portals). It's also a good example of micro-optims at the cost of losing semantics: some assumptions don't always hold, and problems happen. I'd revert that part.

Reply:
Yup, was experimenting to see if it would break tests. Will revert it.

Reply:
I temporarily changed it to a better implementation that is still significantly faster.

Before:
![image](https://user-images.githubusercontent.com/697707/99388985-0a445300-28df-11eb-8d1e-116747476421.png)

After:
![image](https://user-images.githubusercontent.com/697707/99388958-03b5db80-28df-11eb-9553-61cc87f35f77.png)

This would mostly show improvements in bigger monorepos, such as the one that was discussed on Discord the other day.

Reply:
It's still the very last change I would merge.

Reply:
Note that `ppath.join` shows up as all 3 of `join` `normalize` and `normalizeString` in the top screenshot, because of how it is implemented.

So the total time of 14ms vs 5ms for `hydrateRuntimeState` is the relevant figure.

Reply:
I know 🙂